src/Benchmarks/Middleware/JsonMiddleware.cs

```diff
@@ -4,34 +4,53 @@
 using System;
 using System.IO;
 using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
 using System.Threading.Tasks;
 using Benchmarks.Configuration;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http;
-using Microsoft.AspNetCore.Http.Features;
-using System.Text.Json;
-using System.Text.Json.Serialization;
+using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Http.Json;
+using System.Text.Json.Serialization.Metadata;
 
 namespace Benchmarks.Middleware
 {
     public class JsonMiddleware
     {
         private static readonly PathString _path = new PathString(Scenarios.GetPath(s => s.Json));
-        private static readonly UTF8Encoding _encoding = new UTF8Encoding(false);
         private const int _bufferSize = 27;
         private readonly RequestDelegate _next;
+#if NET8_0_OR_GREATER
+        private readonly JsonTypeInfo<JsonMessage> _jsonTypeInfo;
+#else
+        private readonly JsonSerializerOptions _jsonOptions;
+#endif
 
-        public JsonMiddleware(RequestDelegate next) => _next = next;
+        public JsonMiddleware(RequestDelegate next, IOptions<JsonOptions> jsonOptions)
+        {
+            _next = next;
+#if NET8_0_OR_GREATER
+            _jsonTypeInfo = JsonTypeInfo.CreateJsonTypeInfo<JsonMessage>(jsonOptions.Value.SerializerOptions);
```

Comment:
@eiriktsarpalis is `JsonTypeInfo.CreateJsonTypeInfo<JsonMessage>(jsonOptions.Value.SerializerOptions);` supposed to create a `JsonTypeInfo<T>` that doesn't seem to work?

Reply:
That just creates a blank, mutable `JsonTypeInfo` for the given type. You should use `(JsonTypeInfo<JsonMessage>)jsonOptions.Value.SerializerOptions.GetTypeInfo(typeof(JsonMessage));` instead.